frontend/lib/resolution/Resolver.cpp

```diff
@@ -2992,10 +2992,25 @@ void Resolver::resolveIdentifier(const Identifier* ident) {
   }
 
   // lookupIdentifier reports any errors that are needed
+  bool resolvingCalledIdent = nearestCalledExpression() == ident;
   auto parenlessInfo = ParenlessOverloadInfo();
-  auto ids = lookupIdentifier(ident, parenlessInfo);
+  auto ids = lookupIdentifier(ident, resolvingCalledIdent, parenlessInfo);
+
+  // If we looked up a called identifier and found ambiguity between variables
+  // only, resolve as an implicit 'this' call on the innermost variable.
+  if (resolvingCalledIdent && ids.numIds() > 1) {
```

Comment:
This works for me in the near term but I view it as a workaround. Could you add a TODO comment here to investigate making the scope lookup code handle this case? It should be getting `LOOKUP_INNERMOST` and as a result it should stop when it finds the variable, which should happen before it reaches the method scope & looks for fields/methods.

Reply:
Will add a TODO.

About the longer term solution though, when should we set `LOOKUP_INNERMOST`? We don't know until we do the lookup whether there will be any proc results, and we can only use it if there aren't.

Reply:
If I'm understanding you correctly, that's the status quo which had the buggy behavior for `this` calls.

Reply:
Yes, that's right.

Reply:
Could you explain what you meant by this then? I thought it was a suggestion for what to change for a more principled solution in the future:
> It should be getting LOOKUP_INNERMOST and as a result it should stop when it finds the variable, which should happen before it reaches the method scope & looks for fields/methods.

Reply:
I'm not sure where the disconnect in understanding each other here is -- but I don't think it matters greatly. Some sort of TODO comment is sufficient.

Nonetheless, I'm saying that I think the bug needs to be fixed in the `LookupHelper` methods that implement scope resolution, and that the code calling into it here was already OK. Of course, I could be wrong about this (and I have not tried to fix this bug like you have).